api/v1alpha1/awsloadbalancercontroller_types.go

```diff
@@ -20,22 +20,85 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+// +kubebuilder:validation:Enum=AWSShield;AWSAddonWAFv1;AWSAddonWAFv2
+type AWSAddon string
+
+const (
+	AWSAddonShield AWSAddon = "AWSShield"
+	AWSAddonWAFv1  AWSAddon = "AWSAddonWAFv1"
+	AWSAddonWAFv2  AWSAddon = "AWSAddonWAFv2"
+)
+
+// +kubebuilder:validation:Enum=Auto;Manual
+type SubnetTaggingPolicy string
+
+const (
+
+	// AutoSubnetTaggingPolicy enables automatic subnet tagging.
+	AutoSubnetTaggingPolicy SubnetTaggingPolicy = "Auto"
+
+	// ManualSubnetTaggingPolicy disables automatic subnet tagging.
+	ManualSubnetTaggingPolicy SubnetTaggingPolicy = "Manual"
+)
 
 // AWSLoadBalancerControllerSpec defines the desired state of AWSLoadBalancerController
 type AWSLoadBalancerControllerSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of AWSLoadBalancerController. Edit awsloadbalancercontroller_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// SubnetTagging describes how resource tagging will be done by the operator.
+	//
+	// When in "Auto", the operator will detect the subnets where the load balancers
+	// will be provisioned and have the required resource tags on them. Whereas when
+	// set to manual, this responsibility lies on the user.
+	//
+	SubnetTagging SubnetTaggingPolicy `json:"subnetTagging,omitempty"`
+
+	// Default AWS Tags that will be applied to all AWS resources managed by this
+	// controller (default []).
+	//
+	// This value is required so that this controller can function as expected
+	// in parallel to openshift-router.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	AdditionalResourceTags map[string]string `json:"additionalResourceTags,omitempty"`
+
+	// IngressClass specifies the Ingress class which the controller will reconcile.
+	// This will default to "alb".
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	IngressClass string `json:"ingressClass,omitempty"`
+
+	// Config specifies further customization options for the controller's deployment spec.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Config AWSLoadBalancerDeploymentConfig `json:"config,omitempty"`
+
+	// AWSAddon describes the AWS services that can be integrated with
+	// the AWS Load Balancer.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	EnabledAddons []AWSAddon `json:"enabledAddons,omitempty"` // indicates which AWS addons should be disabled.
+
 }
 
-// AWSLoadBalancerControllerStatus defines the observed state of AWSLoadBalancerController
+type AWSLoadBalancerDeploymentConfig struct {
+
+	// +kubebuilder:validation:Optional
+	// +optional
+	Replicas int `json:"replicas,omitempty"`
```

Comment:
So, it will be literally not possible to set `replicas` to `0` as it's the default value that will be omitted.

Also, not a big deal but since `AWSLoadBalancerControllerStatus.ObservedGeneration` is set to `int64`, I propose to set to a concrete `int` type this field too: `int32`. This would be more in the way of the API guidelines.

Reply:
Added a default value to replicas as 2 and fixed data type.